src/internal.c

```diff
@@ -6697,11 +6702,441 @@ int SendServiceAccept(WOLFSSH* ssh, byte serviceId)
+
+    if (ret == WS_SUCCESS)
+        ret = wc_HashFinal(&hash, hashId, digest);
+
+    if (ret == WS_SUCCESS) {
+        byte encDigest[MAX_ENCODED_SIG_SZ];
+        int encDigestSz;
+
+        c32toa(keySig->sigSz + 7 + LENGTH_SZ * 2, output + begin);
+        begin += LENGTH_SZ;
+        c32toa(7, output + begin);
+        begin += LENGTH_SZ;
+        WMEMCPY(output + begin, "ssh-rsa", 7);
+        begin += 7;
+        c32toa(keySig->sigSz, output + begin);
+        begin += LENGTH_SZ;
+        encDigestSz = wc_EncodeSignature(encDigest, digest, digestSz,
+                wc_HashGetOID(hashId));
+        if (encDigestSz <= 0) {
+            WLOG(WS_LOG_DEBUG, "SUAR: Bad Encode Sig");
+            ret = WS_CRYPTO_FAILED;
+        }
+        else {
+            int sigSz;
+            WLOG(WS_LOG_INFO, "Signing hash with RSA.");
+            sigSz = wc_RsaSSL_Sign(encDigest, encDigestSz,
+                    output + begin, keySig->sigSz,
+                    &keySig->ks.rsa.key, ssh->rng);
+            if (sigSz <= 0 || (word32)sigSz != keySig->sigSz) {
+                WLOG(WS_LOG_DEBUG, "SUAR: Bad RSA Sign");
+                ret = WS_RSA_E;
+            }
+        }
+
+        if (ret == WS_SUCCESS) {
+            begin += keySig->sigSz;
+            *idx = begin;
+        }
+    }
+
+    return ret;
+}
+
+
+static int PrepareUserAuthRequestEcc(WOLFSSH* ssh, word32* payloadSz,
+        const WS_UserAuthData* authData, WS_KeySignature* keySig)
+{
+    int ret = WS_SUCCESS;
+
+    WLOG(WS_LOG_DEBUG, "Entering PrepareUserAuthRequestEcc()");
+    if (ssh == NULL || payloadSz == NULL || authData == NULL || keySig == NULL)
+        ret = WS_BAD_ARGUMENT;
+
+    if (ret == WS_SUCCESS)
+        ret = wc_ecc_init(&keySig->ks.ecc.key);
+
+    if (ret == WS_SUCCESS) {
+        word32 idx = 0;
+        ret = wc_EccPrivateKeyDecode(authData->sf.publicKey.privateKey,
+                &idx, &keySig->ks.ecc.key,
+                authData->sf.publicKey.privateKeySz);
+    }
+
+    if (ret == WS_SUCCESS) {
+        if (authData->sf.publicKey.hasSignature) {
+            int sigSz = wc_ecc_sig_size(&keySig->ks.ecc.key);
+            sigSz = 64;
+
+            if (sigSz >= 0) {
+                *payloadSz += (LENGTH_SZ * 5) + (word32)sigSz +
+                        authData->sf.publicKey.publicKeyTypeSz;
+                keySig->sigSz = sigSz;
+            }
+            else
+                ret = sigSz;
+        }
+    }
+
+    WLOG(WS_LOG_DEBUG, "Leaving PrepareUserAuthRequestEcc(), ret = %d", ret);
+    return ret;
+}
+
+
+static int BuildUserAuthRequestEcc(WOLFSSH* ssh,
+        byte* output, word32* idx,
+        const WS_UserAuthData* authData,
+        const byte* sigStart, word32 sigStartIdx,
+        WS_KeySignature* keySig)
+{
+    wc_HashAlg hash;
+    byte digest[WC_MAX_DIGEST_SIZE];
+    word32 digestSz;
+    word32 begin;
+    enum wc_HashType hashId = WC_HASH_TYPE_SHA;
+    int ret = WS_SUCCESS;
+    byte* r;
+    byte* s;
+    byte sig[512];
```

Comment:
Should MAX_ECC_BITS be used here instead of 512?

Reply:
It should be big enough to fit a signature for a prime 521 key, so 144 bytes. I think.

Reply:
Reduced the sig buffer to 72 bytes, which is the size of a signature for a prime256 key.